homeassistant/components/here_travel_time/sensor.py

```diff
@@ -0,0 +1,405 @@
+
+    async_add_entities([sensor], True)
+
+
+def _are_valid_client_credentials(here_client: herepy.RoutingApi) -> bool:
+    """Check if the provided credentials are correct using defaults."""
+    test_origin = [38.9, -77.04833]
+    test_destination = [39.0, -77.1]
+    try:
+        here_client.car_route(
+            test_origin,
+            test_destination,
+            [
+                herepy.RouteMode[ROUTE_MODE_FASTEST],
+                herepy.RouteMode[TRAVEL_MODE_CAR],
+                herepy.RouteMode[TRAFFIC_MODE_DISABLED],
+            ],
+        )
+    except herepy.InvalidCredentialsError:
+        return False
+    return True
+
+
+class HERETravelTimeSensor(Entity):
+    """Representation of a HERE travel time sensor."""
+
+    def __init__(
+        self, name: str, origin: str, destination: str, here_data: "HERETravelTimeData"
+    ) -> None:
+        """Initialize the sensor."""
+        self._name = name
+        self._here_data = here_data
+        self._unit_of_measurement = UNIT_OF_MEASUREMENT
+        self._origin_entity_id = None
+        self._destination_entity_id = None
+
+        # Check if location is a trackable entity
+        if origin.split(".", 1)[0] in TRACKABLE_DOMAINS:
+            self._origin_entity_id = origin
+        else:
+            self._here_data.origin = origin
+
+        if destination.split(".", 1)[0] in TRACKABLE_DOMAINS:
+            self._destination_entity_id = destination
+        else:
+            self._here_data.destination = destination
+
+    @property
+    def state(self) -> Optional[str]:
+        """Return the state of the sensor."""
+        if self._here_data.traffic_mode:
+            if self._here_data.traffic_time is not None:
+                return str(round(self._here_data.traffic_time / 60))
+        if self._here_data.base_time is not None:
+            return str(round(self._here_data.base_time / 60))
+
+        return None
+
+    @property
+    def name(self) -> str:
+        """Get the name of the sensor."""
+        return self._name
+
+    @property
+    def device_state_attributes(
+        self
+    ) -> Optional[Dict[str, Union[None, float, str, bool]]]:
+        """Return the state attributes."""
+        if self._here_data.base_time is None:
+            return None
+
+        res = {}
+        res[ATTR_ATTRIBUTION] = self._here_data.attribution
+        res[ATTR_DURATION] = self._here_data.base_time / 60
+        res[ATTR_DISTANCE] = self._here_data.distance
+        res[ATTR_ROUTE] = self._here_data.route
+        res[CONF_UNIT_SYSTEM] = self._here_data.units
+        res[ATTR_DURATION_IN_TRAFFIC] = self._here_data.traffic_time / 60
+        res[ATTR_ORIGIN] = self._here_data.origin
+        res[ATTR_DESTINATION] = self._here_data.destination
+        res[ATTR_ORIGIN_NAME] = self._here_data.origin_name
+        res[ATTR_DESTINATION_NAME] = self._here_data.destination_name
+        res[CONF_MODE] = self._here_data.travel_mode
```

Comment:
Travel mode is static.

We should use `ATTR_*` constants for state attribute keys. `CONF_*` constants are only used for config keys.

Reply:
I changed it to ATTR_MODE will commit it later.

For the static part: I can add it as a parameter as well but then we are back again with many parameters passed to `HERETravelTimeSensor`. I introduced `HERETravelTimeData` because we wanted to avoid that.

Reply:
Or did I get you wrong and you want to do in `__init__` something like:
`self._attr_travel_mode = self._here_data.travel_mode`

Reply:
We can define an instance attribute,in init mee, that is a dict with all state attributes.
```py
self._attrs = {
    self._here_data.travel_mode,
    ...
}
```

Reply:
Do we do that for readibility, performance...? Just want to know so I can learn^^

Reply:
It's more clear to not update static attributes, since they are static. Updating them would give the impression that they are changing. We shouldn't give false impressions.

In general, in some cases, this could have a performance benefit too.